### PR TITLE
Add StableDiffusionXLControlNetPAGImg2ImgPipeline

### LINKED ISSUE #8700: [PAG] add `StableDiffusionXLControlNetPAGImg2ImgPipeline`
We recently integrated PAG into diffusers! See the PR here: https://github.com/huggingface/diffusers/pull/7944

Does anyone want to add a `StableDiffusionXLControlNetPAGImg2ImgPipeline`?
1. You should put it under the [pag folder](https://github.com/huggingface/diffusers/tree/main/src/diffusers/pipelines/pag)
2. you can use the implementation of [`StableDiffusionXLControlNetPAGPipeline`](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_xl.py) and [`StableDiffusionXLPAGImg2ImgPipeline`](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/pag/pipeline_pag_sd_xl_img2img.py) as reference
3. you need to add AutoPipeline so that you can use this API to create it
    ```python
       AutoPipelineForImage2Image.from_pretrained(repo_id, controlnet=controlnet, enable_pag=True ...)
   ```
4. tests and docs

### PR DESCRIPTION
**Pull request by satani99**
# What does this PR do?

fix #8700 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@yiyixuxu

### CODE REVIEW

**Comment by satani99** [comment]
Generation code
```python
import torch
import numpy as np
from PIL import Image 

from transformers import DPTFeatureExtractor, DPTForDepthEstimation
from diffusers import ControlNetModel, AutoencoderKL, AutoPipelineForImage2Image
from diffusers.utils import load_image

depth_estimator = DPTForDepthEstimation.from_pretrained("Intel/dpt-hybrid-midas").to("cuda")
feature_extractor = DPTFeatureExtractor.from_pretrained("Intel/dpt-hybrid-midas")
controlnet = ControlNetModel.from_pretrained(
        "diffusers/controlnet-depth-sdxl-1.0-small",
        variant="fp16",
        use_safetensors="True",
        torch_dtype=torch.float16,
        )
vae = AutoencoderKL.from_pretrained("madebyollin/sdxl-vae-fp16-fix", torch_dtype=torch.float16)
pipe = AutoPipelineForImage2Image.from_pretrained(
        "stabilityai/stable-diffusion-xl-base-1.0",
        controlnet=controlnet,
        vae=vae,
        variant="fp16",
        use_safetensors=True,
        torch_dtype=torch.float16,
        enable_pag=True,
        )
pipe.enable_model_cpu_offload()

def get_depth_map(image):
   image = feature_extractor(images=image, return_tensors="pt").pixel_values.to("cuda")
   with torch.no_grad(), torch.autocast("cuda"):
       depth_map = depth_estimator(image).predicted_depth

   depth_map = torch.nn.functional.interpolate(
        depth_map.unsqueeze(1),
        size=(1024, 1024),
        mode="bicubic",
        align_corners=False,
    )
   depth_min = torch.amin(depth_map, dim=[1, 2, 3], keepdim=True)
   depth_max = torch.amax(depth_map, dim=[1, 2, 3], keepdim=True)
   depth_map = (depth_map - depth_min) / (depth_max - depth_min)
   image = torch.cat([depth_map] * 3, dim=1)
   image = image.permute(0, 2, 3, 1).cpu().numpy()[0]
   image = Image.fromarray((image * 255.0).clip(0, 255).astype(np.uint8))
   return image



prompt = "A robot, 4k photo"
image = load_image(
        "https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main"
        "/kandinsky/cat.png"
        ).resize((1024, 1024))

controlnet_conditioning_scale = 0.5 
depth_image = get_depth_map(image)

images = pipe(
        prompt,
        image=image,
        control_image=depth_image,
        strength=0.99,
        num_inference_steps=50,
        controlnet_conditioning_scale=controlnet_conditioning_scale,
        ).images
images[0].save(f"robot_cat.png")
```
It works with `enable_pag=False` but gives error when `enable_pag=True`.

Error: `AttributeError: 'Image' object has no attribute 'shape'. Did you mean: 'save'?`

**Comment by satani99** [comment]
Any help would be nice. Thanks

**Comment by yiyixuxu** [comment]
can you share the full stack trace?

**Comment by satani99** [comment]
`File "/home/nikhil/Desktop/pag.py", line 72, in <module>
    images = pipe(
  File "/home/nikhil/miniconda3/envs/pag/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
  File "/home/nikhil/Desktop/diffusers/src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_xl_img2img.py", line 1422, in __call__
    height, width = control_image.shape[-2:]
AttributeError: 'Image' object has no attribute 'shape'. Did you mean: 'save'?`
`pag.py ` is the above script.

File: src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_xl_img2img.py [hunk 1]
```diff
@@ -0,0 +1,1703 @@
...
+                device,
+                batch_size * num_images_per_prompt,
+                self.do_classifier_free_guidance,
+            )
+
+        # 4. Prepare image and controlnet_conditioning_image
+        image = self.image_processor.preprocess(image, height=height, width=width).to(dtype=torch.float32)
+
+        if isinstance(controlnet, ControlNetModel):
+            image = self.prepare_control_image(
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/pag/pipeline_pag_controlnet_sd_xl_img2img.py:1411]:
```suggestion
            control_image = self.prepare_control_image(
```

**Comment by satani99** [comment]
hi @yiyixuxu can you review this?

**Comment by yiyixuxu** [comment]
sorry this PR got lost too
could you resolve the conflicts?

File: src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py [hunk 2]
```diff
@@ -1494,8 +1494,8 @@ def __call__(
             dtype=prompt_embeds.dtype,
             text_encoder_projection_dim=text_encoder_projection_dim,
         )
-        add_time_ids = add_time_ids.repeat(batch_size * num_images_per_prompt, 1)
-
+        add_time_ids = add_time_ids.repeat(batch_size * num_images_per_prompt, 1
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/controlnet/pipeline_controlnet_sd_xl_img2img.py:1497]:
```suggestion
        add_time_ids = add_time_ids.repeat(batch_size * num_images_per_prompt, 1)
```

**Comment by yiyixuxu** [comment]
@asomoza 
I tested it, and it works now 
do you want to check if it works as expected for you? (no worries if you don't have time)

**Comment by asomoza** [comment]
@yiyixuxu Tested it and seems ok, it's harder to see the difference here because the base image helps a lot even without PAG, but it still works similar to the other ones.

|w/o pag|with pag|
|---|---|
|||

**Comment by yiyixuxu** [comment]
@satani99 thank you!

**Pull request merged**